In [24]:
import pandas as pd
import missingno as msno 

In [25]:
avocado_df = pd.read_csv(r'C:\Users\abraz\OneDrive\Área de Trabalho\avocado_price_prediction\avocado.csv')
display(avocado_df.head())
display(avocado_df.tail())

Date  AveragePrice  Total Volume     4046      4225    4770  \
0  2015-01-04          1.22      40873.28  2819.50  28287.42   49.90   
1  2015-01-11          1.24      41195.08  1002.85  31640.34  127.12   
2  2015-01-18          1.17      44511.28   914.14  31540.32  135.77   
3  2015-01-25          1.06      45147.50   941.38  33196.16  164.14   
4  2015-02-01          0.99      70873.60  1353.90  60017.20  179.32   

   Total Bags  Small Bags  Large Bags  XLarge Bags          type  year  region  
0     9716.46     9186.93      529.53          0.0  conventional  2015  Albany  
1     8424.77     8036.04      388.73          0.0  conventional  2015  Albany  
2    11921.05    11651.09      269.96          0.0  conventional  2015  Albany  
3    10845.82    10103.35      742.47          0.0  conventional  2015  Albany  
4     9323.18     9170.82      152.36          0.0  conventional  2015  Albany

Date  AveragePrice  Total Volume     4046     4225   4770  \
27318  2015-11-22          1.97       6249.43  1733.40  2873.92  30.95   
27319  2015-11-29          2.08       4638.10  1395.02  2238.04  61.71   
27320  2015-12-13          1.80       7836.65  2194.49  2981.01  25.97   
27321  2015-12-20          1.92       6255.19  1512.45  2407.32  11.78   
27322  2015-12-27          1.81       7155.63  1478.79  2629.64  14.10   

       Total Bags  Small Bags  Large Bags  XLarge Bags     type  year  \
27318     1611.16     1590.00       21.16          0.0  organic  2015   
27319      943.33      943.33        0.00          0.0  organic  2015   
27320     2635.18     2598.45       36.73          0.0  organic  2015   
27321     2323.64     2213.72      109.92          0.0  organic  2015   
27322     3033.10     2855.55      177.55          0.0  organic  2015   

                 region  
27318  WestTexNewMexico  
27319  WestTexNewMexico  
27320  WestTexNewMexico  
27321  WestTexNewMexico  
27322  WestTexNewMexico

In [26]:
avocado_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27323 entries, 0 to 27322
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Date          27323 non-null  object 
 1   AveragePrice  27323 non-null  float64
 2   Total Volume  27323 non-null  float64
 3   4046          27323 non-null  float64
 4   4225          27323 non-null  float64
 5   4770          27323 non-null  float64
 6   Total Bags    27323 non-null  float64
 7   Small Bags    27323 non-null  float64
 8   Large Bags    27323 non-null  float64
 9   XLarge Bags   27323 non-null  float64
 10  type          27323 non-null  object 
 11  year          27323 non-null  int64  
 12  region        27323 non-null  object 
dtypes: float64(9), int64(1), object(3)
memory usage: 2.7+ MB


# Tratamento

Altereando a coluna Date para datetime

In [27]:
avocado_df["Date"] = pd.to_datetime(avocado_df["Date"])
avocado_df.sort_values(by=['Date'], inplace=True, ascending=True)

avocado_df = avocado_df.reset_index(drop=True)

## Dados faltantes

In [28]:
avocado_df.isnull().sum()

Date            0
AveragePrice    0
Total Volume    0
4046            0
4225            0
4770            0
Total Bags      0
Small Bags      0
Large Bags      0
XLarge Bags     0
type            0
year            0
region          0
dtype: int64

## Dados duplicados

In [29]:
avocado_df.duplicated().value_counts()

False    27321
True         2
dtype: int64

In [30]:
avocado_df[avocado_df.duplicated() == True]

Date  AveragePrice  Total Volume     4046      4225   4770  \
12158 2017-02-26          1.16      39054.83  3021.26  15568.68  11.77   
12275 2017-03-05          1.23      24969.30  2292.42   4876.69  52.82   

       Total Bags  Small Bags  Large Bags  XLarge Bags     type  year  \
12158    20453.12    20299.52      153.60          0.0  organic  2017   
12275    17747.37    17114.89      632.48          0.0  organic  2017   

                    region  
12158  West Tex/New Mexico  
12275  West Tex/New Mexico

In [31]:
avocado_df = avocado_df.drop_duplicates()
avocado_df.shape

(27321, 13)

## Aterando nome das colunas

Para facilitar a escrita do código irei substituir o estilo de escrita das colunas para snake_case.

In [32]:
columns = list(avocado_df.columns)

new_columns = [i.replace(" ", '_') for i in columns]

avocado_df.columns = new_columns

#alterando no as colunas 4046, 4225 e 4770 
avocado_df = avocado_df.rename(columns={"4046":"Small_Hass", "4225":"Large_Hass", "4770":"XLarge_Hass"})

avocado_df.head()

Date  AveragePrice  Total_Volume  Small_Hass  Large_Hass  XLarge_Hass  \
0 2015-01-04          1.22      40873.28     2819.50    28287.42        49.90   
1 2015-01-04          1.88      48280.46     4476.73    22389.09        23.15   
2 2015-01-04          1.93      17328.24     2357.18    12692.21         9.47   
3 2015-01-04          1.41       2604.25      220.61       23.64         0.00   
4 2015-01-04          1.19     166006.29    29419.03    47220.75     38568.95   

   Total_Bags  Small_Bags  Large_Bags  XLarge_Bags          type  year  \
0     9716.46     9186.93      529.53          0.0  conventional  2015   
1    21391.49    21379.79       11.70          0.0       organic  2015   
2     2269.38     2269.38        0.00          0.0       organic  2015   
3     2360.00     2360.00        0.00          0.0       organic  2015   
4    50797.56    44329.03     6468.53          0.0  conventional  2015   

             region  
0            Albany  
1         Northeast  
2           NewYork  
3  NewOrleansMobile  
4         Charlotte

## Alterando coluna region

In [33]:
region = list(avocado_df.region.unique())
region

['Albany',
 'Northeast',
 'NewYork',
 'NewOrleansMobile',
 'Charlotte',
 'Nashville',
 'Midsouth',
 'MiamiFtLauderdale',
 'Louisville',
 'Chicago',
 'LosAngeles',
 'NorthernNewEngland',
 'LasVegas',
 'Indianapolis',
 'Houston',
 'HartfordSpringfield',
 'HarrisburgScranton',
 'CincinnatiDayton',
 'Columbus',
 'GreatLakes',
 'GrandRapids',
 'Detroit',
 'Denver',
 'Jacksonville',
 'California',
 'BuffaloRochester',
 'Orlando',
 'Atlanta',
 'Tampa',
 'Syracuse',
 'StLouis',
 'Spokane',
 'Southeast',
 'SouthCentral',
 'SouthCarolina',
 'Seattle',
 'BaltimoreWashington',
 'Boise',
 'SanFrancisco',
 'SanDiego',
 'Sacramento',
 'Roanoke',
 'Boston',
 'RichmondNorfolk',
 'RaleighGreensboro',
 'Portland',
 'Plains',
 'Pittsburgh',
 'PhoenixTucson',
 'Philadelphia',
 'DallasFtWorth',
 'TotalUS',
 'WestTexNewMexico',
 'West',
 'Miami/Ft. Lauderdale',
 'Los Angeles',
 'Las Vegas',
 'South Carolina',
 'Cincinnati/Dayton',
 'Hartford/Springfield',
 'Harrisburg/Scranton',
 'South Central',
 'San Diego

In [34]:
avocado_df.region.value_counts()

Albany               506
Seattle              506
Jacksonville         506
California           506
West                 506
                    ... 
RichmondNorfolk      104
RaleighGreensboro    104
PhoenixTucson        104
DallasFtWorth        104
WestTexNewMexico     103
Name: region, Length: 78, dtype: int64

Existe um erro no preenchimento dos valores da coluna region. Os valores das regiões deveriam somar 504 porém, para os valores que tem espaço  ou barra em sua estrutura (por exemplo Los Angeles ou Baltimore/Washington), foram dividido em dois grupos um com o nome escrito com espaço/barra com um total de 402 valores, e ou com o nome concatenado (exemplo: LosAngeles e BaltimoreWashington) com um total de 104 valores.

In [35]:
def Alteralinha(col):
    linha = col
    linha = col.replace(' ','')
    linha = linha.replace('.', '')
    linha = linha.replace('/', '')
    return linha

In [36]:
avocado_df.region = avocado_df.region.apply(lambda x: Alteralinha(x))
avocado_df.region.value_counts()

Albany                 506
Northeast              506
Syracuse               506
StLouis                506
Spokane                506
Southeast              506
SouthCentral           506
SouthCarolina          506
Seattle                506
BaltimoreWashington    506
Boise                  506
SanFrancisco           506
SanDiego               506
Sacramento             506
Roanoke                506
Boston                 506
RichmondNorfolk        506
RaleighGreensboro      506
Portland               506
Plains                 506
Pittsburgh             506
PhoenixTucson          506
Philadelphia           506
DallasFtWorth          506
TotalUS                506
Tampa                  506
Atlanta                506
Orlando                506
BuffaloRochester       506
NewYork                506
NewOrleansMobile       506
Charlotte              506
Nashville              506
Midsouth               506
MiamiFtLauderdale      506
Louisville             506
Chicago                506
L

In [37]:
avocado_df.region.unique()

array(['Albany', 'Northeast', 'NewYork', 'NewOrleansMobile', 'Charlotte',
       'Nashville', 'Midsouth', 'MiamiFtLauderdale', 'Louisville',
       'Chicago', 'LosAngeles', 'NorthernNewEngland', 'LasVegas',
       'Indianapolis', 'Houston', 'HartfordSpringfield',
       'HarrisburgScranton', 'CincinnatiDayton', 'Columbus', 'GreatLakes',
       'GrandRapids', 'Detroit', 'Denver', 'Jacksonville', 'California',
       'BuffaloRochester', 'Orlando', 'Atlanta', 'Tampa', 'Syracuse',
       'StLouis', 'Spokane', 'Southeast', 'SouthCentral', 'SouthCarolina',
       'Seattle', 'BaltimoreWashington', 'Boise', 'SanFrancisco',
       'SanDiego', 'Sacramento', 'Roanoke', 'Boston', 'RichmondNorfolk',
       'RaleighGreensboro', 'Portland', 'Plains', 'Pittsburgh',
       'PhoenixTucson', 'Philadelphia', 'DallasFtWorth', 'TotalUS',
       'WestTexNewMexico', 'West'], dtype=object)

Irei remover os dados dos estados para evitar a multicolineariedade.

In [38]:
regions_to_remove = ['California', 'GreatLakes', 'Midsouth', 'NewYork', 'Northeast', 'SouthCarolina', 'Plains', 'SouthCentral', 'Southeast', 'TotalUS', 'West']
avocado_df = avocado_df[~avocado_df.region.isin(regions_to_remove)]
avocado_df

Date  AveragePrice  Total_Volume  Small_Hass  Large_Hass  \
0     2015-01-04          1.22      40873.28     2819.50    28287.42   
3     2015-01-04          1.41       2604.25      220.61       23.64   
4     2015-01-04          1.19     166006.29    29419.03    47220.75   
5     2015-01-04          1.84       3966.00      244.34     2700.02   
7     2015-01-04          1.82       2288.44     1438.04       53.73   
...          ...           ...           ...         ...         ...   
27316 2019-12-01          1.20     206981.00    81301.00    35823.00   
27317 2019-12-01          1.56      84384.00     8866.00     2414.00   
27319 2019-12-01          1.80       9482.00     1112.00      614.00   
27321 2019-12-01          1.15      88296.00     3121.00    70033.00   
27322 2019-12-01          2.00       8389.00        7.00     2728.00   

       XLarge_Hass  Total_Bags  Small_Bags  Large_Bags  XLarge_Bags  \
0            49.90     9716.46     9186.93      529.53          0.0   
3             0.00     2360.00     2360.00        0.00          0.0   
4         38568.95    50797.56    44329.03     6468.53          0.0   
5            76.21      945.43      838.34      107.09          0.0   
7             0.00      796.67      796.67        0.00          0.0   
...            ...         ...         ...         ...          ...   
27316      1844.00    88013.00    74221.00    13695.00         97.0   
27317         0.00    73104.00    72648.00      456.00          0.0   
27319         0.00     7755.00     7251.00      505.00          0.0   
27321        77.00    15065.00     7087.00     7212.00        766.0   
27322        37.00     5617.00     5614.00        3.00          0.0   

               type  year             region  
0      conventional  2015             Albany  
3           organic  2015   NewOrleansMobile  
4      conventional  2015          Charlotte  
5           organic  2015          Nashville  
7           organic  2015  MiamiFtLauderdale  
...             ...   ...                ...  
27316  conventional  2019          Charlotte  
27317       organic  2019         LosAngeles  
27319       organic  2019           LasVegas  
27321  conventional  2019             Albany  
27322       organic  2019          Charlotte  

[21755 rows x 13 columns]

In [39]:
avocado_df.head()

Date  AveragePrice  Total_Volume  Small_Hass  Large_Hass  XLarge_Hass  \
0 2015-01-04          1.22      40873.28     2819.50    28287.42        49.90   
3 2015-01-04          1.41       2604.25      220.61       23.64         0.00   
4 2015-01-04          1.19     166006.29    29419.03    47220.75     38568.95   
5 2015-01-04          1.84       3966.00      244.34     2700.02        76.21   
7 2015-01-04          1.82       2288.44     1438.04       53.73         0.00   

   Total_Bags  Small_Bags  Large_Bags  XLarge_Bags          type  year  \
0     9716.46     9186.93      529.53          0.0  conventional  2015   
3     2360.00     2360.00        0.00          0.0       organic  2015   
4    50797.56    44329.03     6468.53          0.0  conventional  2015   
5      945.43      838.34      107.09          0.0       organic  2015   
7      796.67      796.67        0.00          0.0       organic  2015   

              region  
0             Albany  
3   NewOrleansMobile  
4          Charlotte  
5          Nashville  
7  MiamiFtLauderdale

In [43]:
avocado_df.to_csv('avocados_tratado.csv')